enumerable.rb

```diff
@@ -0,0 +1,189 @@
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    pos = 0
+    arr = to_a
+    while pos < arr.length
+      yield(arr[pos])
+      pos += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    arr = to_a
+    pos = 0
+    my_arr = []
+    while pos < arr.length
+      yield(arr[pos], pos)
+      my_arr << arr[pos]
+      my_arr << pos
+      pos += 1
+    end
+    my_arr
+  end

```

Comment:
- [ ] when a block is given when self is an array it should return the array itself after calling the given block once for each element in self

```Ruby
  array = Array.new(100) { rand(0...9) }
  block = proc { |num| num < (0+ 9) / 2 }

  array.my_each_with_index(&block)
  # should be [5, 7, 1, 4, 1, 2, 0, 7, 7, 5, 5, 5, 0, 4, 0, 6, 6, 1, 6, 6, 4, 4, 3, 2, 4, 2, 3, 7, 0, 1, 4, 2, 8, 2..., 5, 4, 8, 1, 4, 6, 3, 8, 7, 2, 8, 2, 3, 3, 2, 6, 1, 7, 6, 6, 8, 2, 4, 2, 5, 1, 5, 0, 4, 6, 7, 3, 7]
  # but got [5, 0, 7, 1, 1, 2, 4, 3, 1, 4, 2, 5, 0, 6, 7, 7, 7, 8, 5, 9, 5, 10, 5, 11, 0, 12, 4, 13, 0, 14, 6, 15...5, 6, 86, 8, 87, 2, 88, 4, 89, 2, 90, 5, 91, 1, 92, 5, 93, 0, 94, 4, 95, 6, 96, 7, 97, 3, 98, 7, 99] instead

```

- [ ] when a block is given when self is a range it should return the range itself after calling the given block once for each element in self


```Ruby
  block = proc { |num| num < (0+ 9) / 2 }
  ange = Range.new(5, 50)

  range.my_each_with_index(&block)
  # should be 5..50
  # but got [5, 0, 6, 1, 7, 2, 8, 3, 9, 4, 10, 5, 11, 6, 12, 7, 13, 8, 14, 9, 15, 10, 16, 11, 17, 12, 18, 13, 19,... 33, 39, 34, 40, 35, 41, 36, 42, 37, 43, 38, 44, 39, 45, 40, 46, 41, 47, 42, 48, 43, 49, 44, 50, 45] instead

```

Reply:
## Please kindly clarify
 ( array = Array.new(100) { rand(0...9) } Random number generated so they will differ
  block = proc { |num| num < (0+ 9) / 2 }

  array.my_each_with_index(&block)
   should be [5, 7, 1, 4, 1, 2, 0, 7, 7, 5, 5, 5, 0, 4, 0, 6, 6, 1, 6, 6, 4, 4, 3, 2, 4, 2, 3, 7, 0, 1, 4, 2, 8, 2..., 5, 4, 8, 1, 4, 6, 3, 8, 7, 2, 8, 2, 3, 3, 2, 6, 1, 7, 6, 6, 8, 2, 4, 2, 5, 1, 5, 0, 4, 6, 7, 3, 7]
   but got [5, 0, 7, 1, 1, 2, 4, 3, 1, 4, 2, 5, 0, 6, 7, 7)

Reply:
hey, @tnyandoro, we compare the result using #my_each_with_index and #each_with_index using the same variables, so even if the number is randomly generated, it should give the same result :)